doc/specs/#1502 - Advanced Tab Switcher/spec.md

```diff
@@ -0,0 +1,219 @@
+
+## Capabilities
+
+### Accessibility
+
+- The tab switcher will be using WinUI, and so it'll be automatically linked to the UIA tree. This allows screen readers to find it, and so narrator will be able to navigate the switcher easily.
+- The UI is also fully keyboard-driven, with the option of using a mouse to interact with the UI.
+- When the tab switcher pops up, the focus immediately swaps to it.
+- For the sake of contrast with the background, we could make it so that it's similar to how the About Page is presented.
+Specifically, a box would pop up, and the rest of the terminal would turn white with some transparency so that the focus is very clearly on the tab switcher.
+
+### Security
+
+This shouldn't introduce any security issues.
+
+### Reliability
+
+How we're updating the MRU is something to watch out for since it triggers on a lot of tab interactions. However, I don't foresee the update taking long at all, and I can't imagine that users can create and delete tabs fast enough to matter.
+
+### Compatibility
+
+- The existing way of navigating horizontally through the tabs on the tab bar should not break.
+- These should also be separate keybindings from the keybindings associated with using the tab switcher.
+- Reordering tabs on the tab bar shouldn't change the MRU order. The MRU is essentially supposed to be a different _view_ of the already existing tab list.
+
+### Performance, Power, and Efficiency
+
+## Potential Issues
+
+We'll need to be careful about how the UI is presented depending on different sizes of the terminal. We also should test how the UI looks as it's open and resizing is happening. Visual Studio's tab switcher is a fixed size, and is always in the middle. Even when the VS window is smaller than the tab switcher size, the tab switcher will show up larger than the VS window itself.
+
+![Small Visual Studio Without Tab Switcher](img/VSMinimumSize.png)
+![Small Visual Studio With Tab Switcher](img/VSMinimumSizeWithTabSwitcher.png)
+
+Visual Studio Code only allows the user to shrink the window until it hits a minimum width and height. This minimum width and height gives its tab switcher enough space to show a meaningful amount of information.
+
+![Small Visual Studio Code with Tab Switcher](img/VSCodeMinimumTabSwitcherSize.png)
+
+Terminal can't really replicate Visual Studio's version of the tab switcher in this situation. The TabSwitcher needs to be contained within the Terminal. So, if the TabSwitcher is always centered and has a percentage padding from the borders of the Terminal, it'll shrink as Terminal shrinks. One thing I'm not too sure about is when the Terminal window is so small that not even a single tab title will be visible. Should it just not display the UI? Should it try to display the UI and whatever is visible, is visible?

```

Comment:
So, if we don't have enough real estate on the screen to render a tab switcher, we'll render it but you won't be able to see it well? Why not create some kind of sound or visual indication that we failed to load it?

I'm thinking about users that can't see the Terminal. They would need some kind of notification that the behavior has changed, or nothing happened.

Reply:
The switcher won't be really visible at super small sizes of the terminal, but it'll still activate and will be in focus and will accept keybindings like normal. The user might not see anything happen (because the terminal is super small anyway), but the functionality is still there. It shouldn't fail to load.

So, for the users that can't see the Terminal, they'll still be able to navigate through the switcher using the keyboard, and the narrator should be able to read out each tab as the user switches through the tabs.